Database Init, K3s, NATS, Grouper, plus fixes

johnworth opened the pull request:
I originally intended for this to only encompass the k3s installation and configuration, but that task unwound into a bunch of related work.

I've provided playbooks for the following tasks:
* Initializing the databases, including `k3s` and `de_releases`.
* Uninstalling a k8s cluster.
* Installing a k3s cluster.
* Deploying and configuring NATS inside the cluster.
* Initializing the grouper database.

Commit:
Add de_releases database support

Commit:
Begin putting together everything needed for the k3s install.

Commit:
More work on installing a k3s cluster backed by postgresql

Commit:
Add playbook to uninstall k8s (not k3s)

Commit:
Set pg_standard_conforming_string to on

Commit:
Move init.yaml vars into group_vars, use command module for migrate execs, use first host in dbms_primary group for login_host

Commit:
In main.yml, enable fact gathering, add setting up pg_hba.conf, open port 5432, and install acl package

Commit:
Add more sane defaults to group_vars/all.yml for postgresql

Commit:
Reorged install_k3s playbook to reduce fact gathering. Fixed a bunch of plays.

Commit:
Removed k8s specific roles and playbooks, renamed haproxy role, added haproxy playbook, fixed up reverse proxy support in k3s configuration

Commit:
Added ability to force a k3s reinstall, added configurable version to install

Commit:
Add playbook for doing system updates

Commit:
Added playbook for initializing grouper

Commit:
Fix grouper database ownership

Commit:
Update the cert-manager custom file

Commit:
Adjust timeouts for the k8s commands in the grouper playbook. Add debug output

Commit:
Add more docs and a few fixes that I ran across when reviewing playbooks

Commit:
Move the vars from the nats/install.yml playbook into the nats/group_vars/all.yml file

Commit:
Added more documentation to README.md

Commit:
Fixed a typo in README.md

slr71 commented on ansible/README.md:
```diff
@@ -1,10 +1,42 @@
+
+The owner users are configurable through the `dbms_connection_user` and `grouper_connection_user` group_vars.
+
+Migrations are run for the `de`, `metadata`, `notifications`, and `de_releases` databases. The `k3s` database is initialized by the installation process for the k3s cluster and the `grouper` database is handled by it's own playbook since it's fairly complicated. The `qms`
```
There's a typo on this line.

```suggestion 
Migrations are run for the `de`, `metadata`, `notifications`, and `de_releases` databases. The `k3s` database is initialized by the installation process for the k3s cluster and the `grouper` database is handled by its own playbook since it's fairly complicated. The `qms`
```

slr71 commented on ansible/README.md:
```diff
@@ -1,10 +1,42 @@
+The owner users are configurable through the `dbms_connection_user` and `grouper_connection_user` group_vars.
+
+Migrations are run for the `de`, `metadata`, `notifications`, and `de_releases` databases. The `k3s` database is initialized by the installation process for the k3s cluster and the `grouper` database is handled by it's own playbook since it's fairly complicated. The `qms`
+database is created here, but populated by the `qms` service. `unleash` is not yet initialized by this playbook.
```
QMS is currently configured not to apply migrations to the `qms` database by default, so it might be useful to have this playbook run the QMS migrations as well. These migrations are currently in the QMS repository, but I'd be open to moving them if that makes life easier.

johnworth replied:
Will do, thanks for the heads up.

johnworth replied:
This is fixed.

slr71 commented on ansible/README.md:
```diff
@@ -1,10 +1,42 @@
+
+| Playbook            | Description                                    | Example                                                                                   |
+| ------------------- | ---------------------------------------------- | ----------------------------------------------------------------------------------------- |
+| postgresql/init.yml | Creates the databases and runs some migrations | `ansible-playbook -i <inventory> -e "@</path/to/dbms/group-vars>" -K postgresql/init.yml` |
```
Would it be possible to set `hosts` to the DBMS host and use `delegate_to` to have the commands run on the local host? That way, we might be able to eliminate the need to load the group vars separately. I don't know if that will work or not, but I thought I'd ask. 😁

johnworth replied:
I can give that a shot. I seem to remember trying that out, but I tried out a bunch of stuff and may be misremembering things.

johnworth replied:
Turns out I didn't try it out and it worked fine once I remembered to set gather_facts to false. Should be fixed now.

slr71 commented on ansible/README.md:
```diff
@@ -1,10 +1,42 @@
+| ---------------------------- | ---------------------------------- | ----------------------------------------------------------------- |
+| kubernetes/uninstall_k8s.yml | Uninstalls an existing k8s cluster | `ansible-playbook -i <inventory> -K kubernetes/uninstall_k8s.yml` |
+| kubernetes/install_k3s.yml   | Installs k3s                       | `ansible-playbook -i <inventory> -K kubernetes/install_k3s.yml`   |
+| kubernetes/setup_haproxy.yml | Installs k3s reverse proxy         | `ansible-playbook -i <inventory> -K kubernetes/setup_haproxy.yml` |
```
I like having the example commands in these tables. 👍

slr71 commented on ansible/grouper/group_vars/all.yml:
```diff
@@ -0,0 +1,3 @@
+---
+ns: qa
+grouper_gsh_image: harbor.cyverse.org/de/grouper:v3.0.0
```
I interpreted the [variable precedence table](https://docs.ansible.com/ansible/latest/playbook_guide/playbooks_variables.html#understanding-variable-precedence) to mean that playbook group variables have a higher precedence than inventory group variables. This value can still be overridden by specifying inventory group variables for a specific group, but I was afraid that it might cause confusion. Would it be reasonable to move some of these tasks to a role and have the image defined in variable defaults instead?

Also, would it be reasonable to define it like this?

```
grouper_gsh_image: "harbor.cyverse.org/de/grouper:{{grouper_service_version}}"
```

If something like that would work then it would be possible to specify just the service version for sites that are running a different image version.

johnworth replied:
Yeah, I went back and forth on this and couldn't decide. I could do that, or I could do what was done in other playbooks and use `_var_name` in the playbook and allow it to be overridden by a `var_name` variable by setting the value to `"{{ var_name | default('sane default') }}"`. I noticed that the postgresql roles/playbooks do that, which  isn't a bad solution.

I should still be able to split out the version number, either way.

johnworth replied:
I've moved logic for the grouper init into a role and set the variable in the defaults. I also split out the version into a separate var and had change the container image definition to use it in the play.

slr71 commented on ansible/grouper/main.yml:
```diff
@@ -0,0 +1,190 @@
+---
+- name: Initialize grouper
```
🎉

slr71 commented on ansible/grouper/main.yml:
```diff
@@ -0,0 +1,190 @@
+      register: init_result
+      ignore_errors: true
+
+    - name: debug the database init command result
+      debug:
+        var: init_result
+      when: init_result.failed
```
Nice! 👍

slr71 commented on ansible/kubernetes/install_k3s.yml:
```diff
@@ -0,0 +1,186 @@
+        enabled: true
+
+- name: set up remaining k3s control nodes
+  hosts: k3s_controllers[1:]
```
Cool! I didn't know this was possible.

slr71 commented on ansible/kubernetes/update.yml:
```diff
@@ -0,0 +1,47 @@
+---
+- name: system updates
```
Nice!

slr71 commented on ansible/kubernetes/update.yml:
```diff
@@ -0,0 +1,47 @@
+          terminate_grace_period: 0
+      delegate_to: 127.0.0.1
+
+    - name: update packages
+      ansible.builtin.shell: yum update -y --disablerepo=kubernetes
```
This will have to be updated when we migrate to Ubuntu.

johnworth replied:
Yeah, I knocked this together on short notice on a maintenance day and couldn't figure out how to do everything we needed in a cross-distro manner.

johnworth replied:
Without added more plays and vars, I should add. Branching based on OS isn't terribly difficult, just annoying when you're in a hurry.

slr71 commented on ansible/postgresql/init.yml:
```diff
@@ -40,9 +14,20 @@
     - name: init db  | create de database
       community.postgresql.postgresql_db:
-        login_host: "{{ pg_login_host }}"
+        login_host: "{{ groups['dbms_primary'][0] }}"
```
👍

slr71 reviewed (commented):
I had some questions about variable precedence in Ansible, but this looks good!

johnworth committed:
Update ansible/README.md

Co-authored-by: Sarah Roberts <sarahr@cyverse.org>

Commit:
Moves the DB init logic into a the postgresql/roles/init role.

Moves the postgresql default values into a separate role called common, so they can be reused in both the init and postgresql roles.

Commit:
Move grouper init logic into a role with defaults defined in a common role

Commit:
Split out the gsh image version to a separate var

Commit:
Remove the old playbook group_vars/all.yml file

Commit:
In postgresql init role, use delegate_to rather than hosts: localhost

Commit:
Updated the README

Commit:
Add initialization logic for the QMS database to the postgresql/init role
